app/com/linkedin/drelephant/spark/heuristics/StagesWithFailedTasksHeuristic.scala

```diff
@@ -0,0 +1,142 @@
+      new HeuristicResultDetails("Stages with Overhead memory errors", evaluator.stagesWithOverheadError.toString)
+    )
+    if (evaluator.severityOverheadStages.getValue >= Severity.MODERATE.getValue)
+      resultDetails = resultDetails :+ new HeuristicResultDetails("Overhead memory errors", "Some tasks have failed due to overhead memory error. Please try increasing spark.yarn.executor.memoryOverhead by 500MB in spark.yarn.executor.memoryOverhead")
+    //TODO: refine recommendations
+    if (evaluator.severityOOMStages.getValue >= Severity.MODERATE.getValue)
+      resultDetails = resultDetails :+ new HeuristicResultDetails("OOM errors", "Some tasks have failed due to OOM error. Try increasing spark.executor.memory or decreasing spark.memory.fraction (take a look at unified memory heuristic) or decreasing number of cores.")
+    val result = new HeuristicResult(
+      heuristicConfigurationData.getClassName,
+      heuristicConfigurationData.getHeuristicName,
+      evaluator.severity,
+      0,
+      resultDetails.asJava
+    )
+    result
+  }
+}
+
+object StagesWithFailedTasksHeuristic {
+
+  val OOM_ERROR = "java.lang.OutOfMemoryError"
+  val OVERHEAD_MEMORY_ERROR = "killed by YARN for exceeding memory limits"
+  val ratioThreshold : Double = 2
+
+  class Evaluator(memoryFractionHeuristic: StagesWithFailedTasksHeuristic, data: SparkApplicationData) {
+    lazy val stagesWithFailedTasks: Seq[StageData] = data.stagesWithFailedTasks
+
+    /**
+      * @return : returns the OOM and Overhead memory errors severity
+      */
+    private def getErrorsSeverity: (Severity, Severity, Int, Int) = {
+      var severityOOM: Severity = Severity.NONE
+      var severityOverhead: Severity = Severity.NONE
+      var stagesWithOOMError: Int = 0
+      var stagesWithOverheadError: Int = 0
+      stagesWithFailedTasks.foreach(stageData => {
+        val numCompleteTasks: Int = stageData.numCompleteTasks
+        var failedOOMTasks = 0
+        var failedOverheadMemoryTasks = 0
+        stageData.tasks.get.values.foreach((taskData: TaskData) => {
+          var errorMessage: String = taskData.errorMessage.getOrElse("")
+          failedOOMTasks = hasError(errorMessage, OOM_ERROR, failedOOMTasks)
+          failedOverheadMemoryTasks = hasError(errorMessage, OVERHEAD_MEMORY_ERROR, failedOverheadMemoryTasks)
+        })
+        if (failedOOMTasks > 0) {
+          stagesWithOOMError = stagesWithOOMError + 1
+        }
+        if (failedOverheadMemoryTasks > 0) {
+          stagesWithOverheadError = stagesWithOverheadError + 1
+        }
+        severityOOM = getStageSeverity(failedOOMTasks, stageData.status, severityOOM, numCompleteTasks)
+        severityOverhead = getStageSeverity(failedOverheadMemoryTasks, stageData.status, severityOverhead, numCompleteTasks)
+      })
+      (severityOOM, severityOverhead, stagesWithOOMError, stagesWithOverheadError)
+    }
+
+    /**
+      * returns the max (severity of this stage, present severity)
```

Comment:
define what is stage severity here in layman terms first.

Reply:
Done.